cfn-resources/serverless-instance/cmd/resource/resource.go

```diff
@@ -0,0 +1,395 @@
+				OperationStatus:  handler.Failed,
+				Message:          err.Error(),
+				HandlerErrorCode: cloudformation.HandlerErrorCodeAlreadyExists}, nil
+		}
+		return progress_events.GetFailedEventByResponse(err.Error(), res.Response), nil
+	}
+
+	return handler.ProgressEvent{
+		OperationStatus:      handler.InProgress,
+		Message:              fmt.Sprintf("Create ServerlessInstance `%s`", serverless.StateName),
+		ResourceModel:        currentModel,
+		CallbackDelaySeconds: CallBackSeconds,
+		CallbackContext: map[string]interface{}{
+			constants.StateName: serverless.StateName,
+		},
+	}, nil
+}
+
+func Read(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
+	setup()
+
+	_, _ = log.Debugf("Read() currentModel:%+v", currentModel)
+
+	// Validation
+	modelValidation := validateModel(ReadRequiredFields, currentModel)
+	if modelValidation != nil {
+		return *modelValidation, nil
+	}
+
+	// Create atlas client
+	client, err := util.CreateMongoDBClient(*currentModel.ApiKeys.PublicKey, *currentModel.ApiKeys.PrivateKey)
+	if err != nil {
+		_, _ = log.Warnf("Read - error: %+v", err)
+		return progress_events.GetFailedEventByCode(fmt.Sprintf("Error creating mongoDB client : %s", err.Error()),
+			cloudformation.HandlerErrorCodeInvalidRequest), nil
+	}
+
+	cluster, res, err := client.ServerlessInstances.Get(context.Background(), *currentModel.ProjectID, *currentModel.Name)
+	if err != nil {
+		_, _ = log.Warnf("Read - error: %+v", err)
+		return progress_events.GetFailedEventByResponse(err.Error(), res.Response), nil
+	}
+	// Read Instance
+	model := readServerlessInstance(cluster)
+	model.ApiKeys = &ApiKeyDefinition{
+		PrivateKey: currentModel.ApiKeys.PrivateKey,
+		PublicKey:  currentModel.ApiKeys.PublicKey,
+	}
+	// Response
+	return handler.ProgressEvent{
+		OperationStatus: handler.Success,
+		ResourceModel:   model,
+	}, nil
+}
+
+func Update(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
+	setup()
+
+	_, _ = log.Debugf("Update() currentModel:%+v", currentModel)
+
+	// Validation
+	modelValidation := validateModel(UpdateRequiredFields, currentModel)
+	if modelValidation != nil {
+		return *modelValidation, nil
+	}
+
+	// Create atlas client
+	client, err := util.CreateMongoDBClient(*currentModel.ApiKeys.PublicKey, *currentModel.ApiKeys.PrivateKey)
+	if err != nil {
+		_, _ = log.Warnf("Update - error: %+v", err)
+		return progress_events.GetFailedEventByCode(fmt.Sprintf("Error creating mongoDB client : %s", err.Error()),
+			cloudformation.HandlerErrorCodeInvalidRequest), nil
+	}
+
+	// Callback
+	if _, ok := req.CallbackContext[constants.StateName]; ok {
+		return serverlessCallback(client, currentModel, constants.IdleState)
+	}
+
+	// CFN TEST : currently Update is throwing 500 Error instead of 404 if resource not exists
+	_, res, err := client.ServerlessInstances.Get(context.Background(), *currentModel.ProjectID, *currentModel.Name)
```

Comment:
can we validate the 500 error instead of doing a get to fix the error?

Reply:
500 error is from API we can't fix it , and don't want to depend on API changes